src/common/utils.cc

```diff
@@ -41,20 +62,22 @@ NormalizeVec(float* x, int32_t d) {
 }
 
 // normalize all vectors and return their norms
+template <typename DataType>
 std::vector<float>
-NormalizeVecs(float* x, size_t rows, int32_t dim) {
+NormalizeVecs(DataType* x, size_t rows, int32_t dim) {
     std::vector<float> norms(rows);
     for (size_t i = 0; i < rows; i++) {
         norms[i] = NormalizeVec(x + i * dim, dim);
```

Comment:
should be "norms[i] = NormalizeVec<DataType>(x + i * dim, dim);" ?

Reply:
The compiler can infer the data type by x.